src/java/org/apache/cassandra/index/sai/plan/QueryView.java

```diff
@@ -101,80 +103,156 @@ static class Builder
             this.queryContext = queryContext;
         }
 
+        /**
+         * Denotes a situation when there exist no index for an active memtable or sstable.
+         * This can happen e.g. when the index gets dropped while running the query.
+         */
+        static class MissingIndexException extends RuntimeException
+        {
+            final IndexContext context;
+
+            public MissingIndexException(IndexContext context, String message)
+            {
+                super(message);
+                this.context = context;
+            }
+        }
+
         /**
          * Acquire references to all the memtables, memtable indexes, sstables, and sstable indexes required for the
          * given expression.
-         * <p>
-         * Will retry if the active sstables change concurrently.
          */
-        protected QueryView build()
+        protected QueryView build() throws MissingIndexException
         {
             var referencedIndexes = new HashSet<SSTableIndex>();
-            long failingSince = -1L;
+
+            // We must use the canonical view in order for the equality check for source sstable/memtable
+            // to work correctly.
+            var filter = RangeUtil.coversFullRing(range)
+                         ? View.selectFunction(SSTableSet.CANONICAL)
+                         : View.select(SSTableSet.CANONICAL, s -> RangeUtil.intersects(s, range));
+
+
             try
             {
+                // Keeps track of which memtables we've already tried to match the index to.
+                // If we fail to match the index to the memtable for the first time, we have to retry
+                // because the memtable could be flushed and its index removed between the moment we
+                // got the view and the moment we did the lookup.
+                // If we get the same memtable in the view again, and there is no index,
+                // then the missing index is not due to a concurrent modification, but it doesn't contain indexed
+                // data, so we can ignore it.
```

Comment:
Can you justify this point a bit more? It seems dependent on how long it takes to retry getting the view.